include/pstl_offload/internal/usm_memory_replacement.h

```diff
@@ -53,34 +55,57 @@ __get_offload_device()
 class __offload_policy_holder_type
 {
     using __set_active_device_func_type = void (*)(sycl::device*);
+    using __get_offload_device_func_type = sycl::device (*)();
 
   public:
     // Since the global object of __offload_policy_holder_type is static but the constructor
     // of the class is inline, we need to avoid calling static functions inside of the constructor
     // and pass the pointer to exact function as an argument to guarantee that the correct __active_device
     // would be stored in each translation unit
-    __offload_policy_holder_type(sycl::device&& __offload_device,
+    __offload_policy_holder_type(__device_selector_type __device_selector,
                                  __set_active_device_func_type __set_active_device_func)
-        : _M_offload_device(std::move(__offload_device)), _M_offload_policy(_M_offload_device),
-          _M_set_active_device(__set_active_device_func)
+        : _M_set_active_device(__set_active_device_func)
     {
-        _M_set_active_device(&_M_offload_device);
+        try
+        {
+            _M_offload_device = sycl::device{__device_selector};
+        }
+        catch (const sycl::exception& e)
+        {
+            // e.code() == sycl::errc::runtime means occurs when unable to get offload device. Do not pass
+            // an exception, as ctor is called for a static object and the exception can't be processed.
+            // Remember the situation and re-throw exception when asked for the policy from user's code.
+            if (e.code() != sycl::errc::runtime)
+                throw;
+        }
+
+        if (_M_offload_device.has_value())
+        {
+            _M_set_active_device(&*_M_offload_device);
```

Comment:
Having `set_active_device` only if the device was constructed will cause the memory replacement function to always use system malloc and the user will identify the error only when the first parallel algorithm would be called. 
I guess it makes sense to store some special value into the active device in the `else` branch here and throw an exception from the allocating functions as well.

Reply:
Interesting point. As you may see from the test, I definitely want to support seamless regression to a non-PSTL offload implementation. I.e., if TSLT offload is unavailable, a user may call original `par`/`seq` algorithm (better yet, non-offload `par_unseq`). But what should she/he use after a `malloc` throws an exception? Call another `malloc` in another TU?

Reply:
As discussed offline, one of the main purposes to have "delayed" throw here is to allow user to fallback into seq/par/unseq version of the algorithm in case of exception from the parallel algorithm. In this case, having an error handling inside allocating functions is not needed because we want allocating functions to provide memory that can be used in seq/par/unseq algorithm calls.